### PR TITLE
 Adds getSlot, getSlotLeader, getminimumLedgerSlot, getGenesisHash, EpochSchedule methods

### PR DESCRIPTION
**Pull request by OmkarAcharekar**
The pr adds 5 json rpc requests of issue #39
* [getSlot](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getSlot)
* [getSlotLeader](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getSlotLeader)
* [getminimumLedgerSlot](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getMinimumLedgerSlot)
* [getGenesisHash](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getGenesisHash)
* [getEpochSchedule](https://solana-labs.github.io/solana-web3.js/classes/Connection.html#getEpochSchedule)

### CODE REVIEW

File: include/solana.hpp [hunk 1]
```diff
@@ -554,6 +579,23 @@ class Connection {
    **/
   Version getVersion() const;
 
+  uint64_t getminimumLedgerSlot() const;
+  // Returns the lowest slot that the node has information about in its ledger.
```
Comment (mschneider) [inline include/solana.hpp:583]: please use block comment before the method declaration so IDEs can pick it up

File: tests/main.cpp [hunk 2]
```diff
@@ -724,3 +724,49 @@ TEST_CASE("getFirstAvailableBlock") {
...
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto slot = connection.getminimumLedgerSlot();
+  CHECK_GT(slot, 0);
+}
+
+TEST_CASE("getGenesisHash") {
+  const solana::Keypair keyPair = solana::Keypair::fromFile(KEY_PAIR_FILE);
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto hash = connection.getGenesisHash();
+  boost::regex expression{
```
Comment (mschneider) [inline tests/main.cpp:758]: genesis hash is basically a constant. no need for regex

File: include/solana.hpp (hunk 1 above)
Comment (OmkarAcharekar) [reply include/solana.hpp]: Done the change..

File: tests/main.cpp (hunk 2 above)
Comment (OmkarAcharekar) [reply tests/main.cpp]: Removed regex check ( now checking that the string should not be empty)

File: tests/main.cpp [hunk 3]
```diff
@@ -724,3 +724,49 @@ TEST_CASE("getFirstAvailableBlock") {
   const auto firstAvailableBlock = connection.getFirstAvailableBlock();
   CHECK_GT(firstAvailableBlock, 0);
 }
+
+TEST_CASE("getSlot") {
+  const solana::Keypair keyPair = solana::Keypair::fromFile(KEY_PAIR_FILE);
```
[the remark this answers is not in the document]
Comment (OmkarAcharekar) [reply tests/main.cpp]: Removed the keypair reading statements

File: tests/main.cpp (hunk 2 above)
Comment (mschneider) [reply tests/main.cpp]: please check for the exact hash, each network has a constant hash. it should not change

File: include/solana.hpp [hunk 4]
```diff
@@ -562,9 +587,35 @@ class Connection {
    **/
   Version getVersion() const;
 
+  /**
+   *  Returns the lowest slot that the node has information about in its ledger.
+   **/
+  uint64_t getminimumLedgerSlot() const;
```
Comment (godmodegalactus) [inline include/solana.hpp:593]: getMinimumLedgerSlot

File: include/solana.hpp [hunk 5]
```diff
@@ -562,9 +587,35 @@ class Connection {
...
+
+  /**
+   * Returns the slot that has reached the given or default commitment level
+   **/
+  uint64_t getSlot(const GetSlotConfig &config = GetSlotConfig{}) const;
+
+  /**
+   * Returns the current slot leader
+   **/
+  std::string getSlotLeader(
```
Comment (godmodegalactus) [inline include/solana.hpp:613]: Why extra return here?

File: lib/solana.cpp [hunk 6]
```diff
@@ -533,9 +550,44 @@ Version Connection::getVersion() const {
...
+  const json params = {config};
+  const json reqJson = jsonRequest("getSlot", params);
+
+  return sendJsonRpcRequest(reqJson);
+}
+
+uint64_t Connection::getminimumLedgerSlot() const {
+  const json params = {};
+
+  const json reqJson = jsonRequest("minimumLedgerSlot", params);
```
Comment (godmodegalactus) [inline lib/solana.cpp:569]: actually should then be named `minimumLedgerSlot`

File: include/solana.hpp [hunk 7]
```diff
@@ -75,6 +75,19 @@ struct Version {
  */
 void from_json(const json &j, Version &version);
 
+struct EpochSchedule {
+  uint64_t firstNormalEpoch;
+  uint64_t firstNormalSlot;
+  uint64_t leaderScheduleSlotOffset;
+  uint64_t slotsPerEpoch;
+  bool warmup;
```
Comment (godmodegalactus) [inline include/solana.hpp:83]: Should implement 
```
    getEpoch(slot: number): number;
    getEpochAndSlotIndex(slot: number): [number, number];
    getFirstSlotInEpoch(epoch: number): number;
    getLastSlotInEpoch(epoch: number): number;
    getSlotsInEpoch(epoch: number): number;
```
as implemented in solana web3. They are essential to follow the schedule. Check solana repo for its implementation.

File: tests/main.cpp [hunk 8]
```diff
@@ -712,8 +711,45 @@ TEST_CASE("getVersion") {
...
   CHECK_GT(firstAvailableBlock, 0);
 }
+
+TEST_CASE("getSlot") {
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto slot = connection.getSlot();
+  CHECK_GT(slot, 0);
+  sleep(2);  // sleep for 2 seconds
+  const auto slot2 = connection.getSlot();
+  CHECK_GE(slot2, slot);
```
Comment (godmodegalactus) [inline tests/main.cpp:725]: SHould be `CHECK_GT`

File: tests/main.cpp [hunk 9]
```diff
@@ -712,8 +711,45 @@ TEST_CASE("getVersion") {
...
+TEST_CASE("minimumLedgerSlot") {
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto slot = connection.getminimumLedgerSlot();
+  CHECK_GT(slot, 0);
+}
+
+TEST_CASE("getGenesisHash") {
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto hash = connection.getGenesisHash();
+  CHECK_GT(hash.length(), 0);
```
Comment (godmodegalactus) [inline tests/main.cpp:744]: Can check with devnet genesis hash. This info is public or you could find it easily using solana cli.

File: tests/main.cpp [hunk 10]
```diff
@@ -712,8 +711,45 @@ TEST_CASE("getVersion") {
...
+}
+
+TEST_CASE("getEpochSchedule") {
+  const auto connection = solana::rpc::Connection(solana::DEVNET);
+  const auto epochschedule = connection.getEpochSchedule();
+  CHECK_GE(epochschedule.firstNormalEpoch, 0);
+  CHECK_GE(epochschedule.firstNormalSlot, 0);
+  CHECK_GE(epochschedule.leaderScheduleSlotOffset, 0);
+  CHECK_GE(epochschedule.slotsPerEpoch, 0);
+  CHECK((epochschedule.warmup == true || epochschedule.warmup == false));
```
Comment (godmodegalactus) [inline tests/main.cpp:754]: This check is meaningless as boolean is either true or false atleast in binary(non quantum) computers.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: tests/main.cpp (hunk 8 above)
Comment (mschneider) [reply tests/main.cpp:725]: better version of this test would be to run getSlot with finalized, confirmed and getSlot with processed commitment. this way the test wouldn't need to wait for 2 seconds and you still get different slots

File: tests/main.cpp (hunk 2 above)
Comment (OmkarAcharekar) [reply tests/main.cpp]: Done

File: include/solana.hpp (hunk 4 above)
Comment (OmkarAcharekar) [reply include/solana.hpp]: Renamed to "minimumLedgerSlot"

File: include/solana.hpp (hunk 5 above)
Comment (OmkarAcharekar) [reply include/solana.hpp]: Removed extra line

File: lib/solana.cpp (hunk 6 above)
Comment (OmkarAcharekar) [reply lib/solana.cpp:591]: renamed to "minimumLedgerSlot"

File: tests/main.cpp (hunk 8 above)
Comment (OmkarAcharekar) [reply tests/main.cpp]: Done... checks `getSlot(finalized) < getSlot(confirmed) < getSlot(processed)`

File: tests/main.cpp (hunk 9 above)
Comment (OmkarAcharekar) [reply tests/main.cpp]: Done

File: tests/main.cpp (hunk 10 above)
Comment (OmkarAcharekar) [reply tests/main.cpp]: Yes. Removed the boolean check

File: include/solana.hpp (hunk 7 above)
Comment (OmkarAcharekar) [reply include/solana.hpp:89]: @godmodegalactus should the tests be similar to[ this]( https://github.com/solana-labs/solana-web3.js/blob/53be88c38dd263df25eac37533b45f300c84e44c/test/epoch-schedule.test.ts#L21) ? Thanks

**Comment by godmodegalactus** [comment]
The test fails also on the main branch.

**Comment by OmkarAcharekar** [comment]
@mschneider  yes..The request airdrop methods throws internal error 
` "error": {
        "code": -32603,
        "message": "Internal error"
    },`

File: include/solana.hpp [hunk 11]
```diff
@@ -75,6 +78,71 @@ struct Version {
...
+    }
+  }
+
+  uint64_t getLastSlotInEpoch(uint64_t epoch) const {
+    return this->getFirstSlotInEpoch(epoch) + this->getSlotsInEpoch(epoch) - 1;
+  }
+
+  uint64_t getSlotsInEpoch(uint64_t epoch) const {
+    if (epoch < this->firstNormalEpoch) {
+      return pow(2, epoch + trailingZeros(MINIMUM_SLOT_PER_EPOCH));
```
Comment (godmodegalactus) [inline include/solana.hpp:134]: this can be optimized using <<

File: include/solana.hpp [hunk 12]
```diff
@@ -75,6 +78,71 @@ struct Version {
...
+      info.push_back(epoch);
+      info.push_back(slotIndex);
+
+      return info;
+    }
+  }
+
+  uint64_t getFirstSlotInEpoch(uint64_t epoch) const {
+    if (epoch <= this->firstNormalEpoch) {
+      return (pow(2, epoch) - 1) * MINIMUM_SLOT_PER_EPOCH;
```
Comment (godmodegalactus) [inline include/solana.hpp:121]: pow(2,n) can be optimized using  << on 1.

File: lib/solana.cpp [hunk 13]
```diff
@@ -70,12 +79,42 @@ Keypair Keypair::fromFile(const std::string &path) {
   return result;
 }
 
+uint64_t trailingZeros(uint64_t n) {
+  int trailingsZeros = 0;
```
Comment (godmodegalactus) [inline lib/solana.cpp:83]: type should be same as return type.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: include/solana.hpp (hunk 11 above)
Comment (OmkarAcharekar) [reply include/solana.hpp]: Yes..Done the change. Thanks

File: include/solana.hpp (hunk 12 above)
Comment (OmkarAcharekar) [reply include/solana.hpp]: Done

File: lib/solana.cpp (hunk 13 above)
Comment (OmkarAcharekar) [reply lib/solana.cpp]: Changed variable trailingsZeros type to` uint64_t`

**Review by godmodegalactus — APPROVED** [review]

**Pull request merged**